src/roots.jl

```diff
@@ -29,20 +53,22 @@ Returns the `RootSearchState` at each iteration. Note: Each iteration mutates
 the `RootSearchState`. Use `copy(state::RootSearchState)` to create an
 independent instance if necessary.
 """
-struct RootSearch{R <: Union{Interval,IntervalBox}, S <: Contractor, T <: Real}
-    region::R
-    contractor::S
+struct RootSearch{V, VR, C <: Contractor, S <: SearchStrategy, T <: Real}
+    state::RootSearchState{V, VR}
+    contractor::C
+    strategy::S
     tolerance::T
 end
 
-eltype{R, T <: RootSearch{R}}(::Type{T}) = RootSearchState{R}
+function RootSearch(region::R, contractor::C, tol::T) where {R <: Union{Interval,IntervalBox}, C <: Contractor, T <: Real}
+    RootSearch(RootSearchState(region), contractor, SearchStrategy(), tol)
+end
+
+eltype{RSS, C, S, T, RS <: RootSearch{RSS, C, S, T}}(::Type{RS}) = RSS
 iteratorsize{T <: RootSearch}(::Type{T}) = Base.SizeUnknown()
 
 function start(iter::RootSearch)
-    state = RootSearchState(iter.region)
-    sizehint!(state.outputs, 100)
-    sizehint!(state.working, 1000)
-    return state
+    return iter.state
```

Comment:
This implementation does not conform to the `Base.Iterator` specification: `start(iter)` should return the initial state of the iterator; not the last state that iterator was in. I can imagine this leading to problems down the line when someone tries to restart a `RootSearch` after it has run for a few iterations. Ideally this should create a fresh state. (Resetting state would probably also lead to some confusion.)